[1/N] Nnapi backend compile and execute

Amyh11325 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62272

### Summary:
Added Android NNAPI delegate implementation of runtime initialization (compilation) and execution.
The delegate's preprocess step was [previously implemented](https://github.com/pytorch/pytorch/pull/62225). Now, the rest of the delegate, which implements client-side execution, is added.

**nnapi_backend_lib.cpp**:
Implementation of delegate's compile and execute. 
`execute()` is essentially a C++ implementation of [`NnapiModule`](https://github.com/pytorch/pytorch/blob/master/torch/backends/_nnapi/prepare.py), which wraps an NNAPI Compilation and handles preparation of weights, inputs, and outputs.
- Any steps that can be done before execution are moved to `compile()`.
    - `init()` cannot be moved to `compile()` because it requires real inputs for dynamic shaping.
    - `shape_compute_module` cannot currently be deserialized in `compile()`, since mobile::Module has no IValue conversion.
- Processed arguments that are modified by `init()` must be kept as member variables. Any other processed arguments are passed through a dictionary, `handles`.

**nnapi_bind.cpp & nnapi_bind.h**: 
Created a header file for `nnapi_bind.cpp`, so that it's NnapiCompilation class can be used by `nnapi_backend_lib.cpp`.
**test_backend_nnapi.py**:
Enabled execution testing.

### Test Plan:
Tested on devserver.
1. Load and unpack a special devserver build of NNAPI: `jf download GICWmAAzUR0eo20TAPasVts8ObhobsIXAAAz --file "nnapi-host-linux.tar.xz"`
2. `export LIBNEURALNETWORKS_PATH=/path/to/libneuralnetworks.so`
3. Run unittests: `python test/test_jit.py TestNnapiBackend` and `python test/test_nnapi.py`

TODO: test with lite interpreter runtime

Reviewers:

Subscribers:

Tasks: T91991928

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

Commit:
Draft: Nnapi backend execute and compile

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "Draft: Nnapi backend execute and compile"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "Draft: Nnapi backend execute and compile"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29944873).

Commit:
Update on "Draft: Nnapi backend execute and compile"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]

Commit:
Update on "Draft: Nnapi backend execute and compile"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]

Commit:
Update on "Draft: Nnapi backend execute and compile"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29944873).

Commit:
Update on "[1/N] Nnapi backend compile and execute"


### Summary:
Added Android NNAPI delegate implementation of runtime initialization (compilation) and execution.
The delegate's preprocess step was [previously implemented](https://github.com/pytorch/pytorch/pull/62225). Now, the reset of the delegate, which implements client-side execution, is added.

**nnapi_backend_lib.cpp**:
Implementation of delegate's compile and execute. 
`execute()` is essentially a C++ implementation of [`NnapiModule`](https://github.com/pytorch/pytorch/blob/master/torch/backends/_nnapi/prepare.py), which wraps an NNAPI Compilation and handles preparation of weights, inputs, and outputs.
- Any steps that can be done before execution are moved to `compile()`.
    - `init()` cannot be moved to `compile()` because it requires real inputs for dynamic shaping.
    - `shape_compute_module` cannot currently be deserialized in `compile()`, since mobile::Module has no IValue conversion.
- Processed arguments that are modified by `init()` must be kept as member variables. Any other processed arguments are passed through a dictionary, `handles`.

**nnapi_bind.cpp & nnapi_bind.h**: 
Created a header file for `nnapi_bind.cpp`, so that it's NnapiCompilation class can be used by `nnapi_backend_lib.cpp`.
**test_backend_nnapi.py**:
Enabled execution testing.

### Test Plan:
Tested on devserver.
1. Load and unpack a special devserver build of NNAPI: `jf download GICWmAAzUR0eo20TAPasVts8ObhobsIXAAAz --file "nnapi-host-linux.tar.xz"`
2. `export LIBNEURALNETWORKS_PATH=/path/to/libneuralnetworks.so`
3. Run unittests: `python test/test_jit.py TestNnapiBackend` and `python test/test_nnapi.py`

TODO: test with lite interpreter runtime

Reviewers:

Subscribers:

Tasks: T91991928

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29944873).

Commit:
Update on "[1/N] Nnapi backend compile and execute"


### Summary:
Added Android NNAPI delegate implementation of runtime initialization (compilation) and execution.
The delegate's preprocess step was [previously implemented](https://github.com/pytorch/pytorch/pull/62225). Now, the reset of the delegate, which implements client-side execution, is added.

**nnapi_backend_lib.cpp**:
Implementation of delegate's compile and execute. 
`execute()` is essentially a C++ implementation of [`NnapiModule`](https://github.com/pytorch/pytorch/blob/master/torch/backends/_nnapi/prepare.py), which wraps an NNAPI Compilation and handles preparation of weights, inputs, and outputs.
- Any steps that can be done before execution are moved to `compile()`.
    - `init()` cannot be moved to `compile()` because it requires real inputs for dynamic shaping.
    - `shape_compute_module` cannot currently be deserialized in `compile()`, since mobile::Module has no IValue conversion.
- Processed arguments that are modified by `init()` must be kept as member variables. Any other processed arguments are passed through a dictionary, `handles`.

**nnapi_bind.cpp & nnapi_bind.h**: 
Created a header file for `nnapi_bind.cpp`, so that it's NnapiCompilation class can be used by `nnapi_backend_lib.cpp`.
**test_backend_nnapi.py**:
Enabled execution testing.

### Test Plan:
Tested on devserver.
1. Load and unpack a special devserver build of NNAPI: `jf download GICWmAAzUR0eo20TAPasVts8ObhobsIXAAAz --file "nnapi-host-linux.tar.xz"`
2. `export LIBNEURALNETWORKS_PATH=/path/to/libneuralnetworks.so`
3. Run unittests: `python test/test_jit.py TestNnapiBackend` and `python test/test_nnapi.py`

TODO: test with lite interpreter runtime

Reviewers:

Subscribers:

Tasks: T91991928

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]

Commit:
Update on "[1/N] Nnapi backend compile and execute"


### Summary:
Added Android NNAPI delegate implementation of runtime initialization (compilation) and execution.
The delegate's preprocess step was [previously implemented](https://github.com/pytorch/pytorch/pull/62225). Now, the reset of the delegate, which implements client-side execution, is added.

**nnapi_backend_lib.cpp**:
Implementation of delegate's compile and execute. 
`execute()` is essentially a C++ implementation of [`NnapiModule`](https://github.com/pytorch/pytorch/blob/master/torch/backends/_nnapi/prepare.py), which wraps an NNAPI Compilation and handles preparation of weights, inputs, and outputs.
- Any steps that can be done before execution are moved to `compile()`.
    - `init()` cannot be moved to `compile()` because it requires real inputs for dynamic shaping.
    - `shape_compute_module` cannot currently be deserialized in `compile()`, since mobile::Module has no IValue conversion.
- Processed arguments that are modified by `init()` must be kept as member variables. Any other processed arguments are passed through a dictionary, `handles`.

**nnapi_bind.cpp & nnapi_bind.h**: 
Created a header file for `nnapi_bind.cpp`, so that it's NnapiCompilation class can be used by `nnapi_backend_lib.cpp`.
**test_backend_nnapi.py**:
Enabled execution testing.

### Test Plan:
Tested on devserver.
1. Load and unpack a special devserver build of NNAPI: `jf download GICWmAAzUR0eo20TAPasVts8ObhobsIXAAAz --file "nnapi-host-linux.tar.xz"`
2. `export LIBNEURALNETWORKS_PATH=/path/to/libneuralnetworks.so`
3. Run unittests: `python test/test_jit.py TestNnapiBackend` and `python test/test_nnapi.py`

TODO: test with lite interpreter runtime

Reviewers:

Subscribers:

Tasks: T91991928

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29944873).

iseeyuan commented on torch/csrc/jit/backends/nnapi/nnapi_backend_lib.cpp:
```diff
@@ -17,23 +28,116 @@ class NnapiBackend : public PyTorchBackendInterface {
+    dict.insert("weights", weights);
+
+    // Save ser_model to member variable
+    ser_model_ = dict.at("ser_model").toTensor();
```
Could you remind me why ser_model has to be a member variable?

Amyh11325 replied:
It was because ser_model is mutated in init(), and since init() is called once by execute(), the only way to carry those changes to later iterations is by making a member variable.
Although now that I'm looking at it again, execute() doesn't actually use ser_model. Removing the member variable might work too. I'll give it a try.

raziel replied:
yeah whatever we can remove from the state would be good.

iseeyuan reviewed (approved):
LGTM. Would be great to have more signals from other reviwers!

raziel commented on torch/csrc/jit/backends/nnapi/nnapi_backend_lib.cpp:
```diff
@@ -17,23 +28,116 @@ class NnapiBackend : public PyTorchBackendInterface {
+    auto dict = processed.toGenericDict();
+
+    // Prepare weights
+    auto weights = dict.at("weights").toTensorList();
+    for (int i = 0; i < weights.size(); i++) {
+      weights.set(i, weights.get(i).contiguous());
+    }
+    dict.insert("weights", weights);
```
This couldn't be done in preprocess because of how values are serialized?

Amyh11325 replied:
Actually, it probably can be done in preprocess, since the weights are in the handles dictionary. Thanks for pointing it out. I can follow up on it with moving ser_model_.

raziel commented on torch/csrc/jit/backends/nnapi/nnapi_backend_lib.cpp:
```diff
@@ -17,23 +28,116 @@ class NnapiBackend : public PyTorchBackendInterface {
+      int fmt = inp_mem_fmts[i];
+      // These constants match the values in DimOrder in serializer.py
+      // TODO: See if it's possible to use those directly.
+      if (fmt == 0) {
```
can we at least put the name of the constants as comments?

raziel reviewed (approved):
Thanks @Amyh11325 

A few comments but looks great!

Amyh11325 commented on aten/src/ATen/nnapi/nnapi_bind.cpp:
```diff
@@ -3,15 +3,18 @@
-namespace {
+namespace bind {
 
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
 nnapi_wrapper* nnapi;
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
 nnapi_wrapper* check_nnapi;
```
These are declared in the header, because the other types declared there (specifically `ModelPtr` and `CompilationPtr`), depend on those global variables.
It's defined in .cpp as well, because if it's directly defined in .h (without `extern`), there are ODR warnings.

raziel commented on aten/src/ATen/nnapi/nnapi_bind.h:
```diff
@@ -0,0 +1,54 @@
+#include <torch/custom_class.h>
+
+#include <ATen/nnapi/nnapi_wrapper.h>
+#include <ATen/nnapi/nnapi_model_loader.h>
```
#include <ATen/nnapi/nnapi_model_loader.h>

raziel commented on aten/src/ATen/nnapi/nnapi_bind.cpp:
```diff
@@ -3,15 +3,18 @@
 #include <ATen/ATen.h>
 #include <torch/custom_class.h>
```
#include <torch/custom_class.h>

raziel reviewed (commented):
Added specific headers I was referring to in my previous comments.

Commit:
Update on "[1/N] Nnapi backend compile and execute"


### Summary:
Added Android NNAPI delegate implementation of runtime initialization (compilation) and execution.
The delegate's preprocess step was [previously implemented](https://github.com/pytorch/pytorch/pull/62225). Now, the rest of the delegate, which implements client-side execution, is added.

**nnapi_backend_lib.cpp**:
Implementation of delegate's compile and execute. 
`execute()` is essentially a C++ implementation of [`NnapiModule`](https://github.com/pytorch/pytorch/blob/master/torch/backends/_nnapi/prepare.py), which wraps an NNAPI Compilation and handles preparation of weights, inputs, and outputs.
- Any steps that can be done before execution are moved to `compile()`.
    - `init()` cannot be moved to `compile()` because it requires real inputs for dynamic shaping.
    - `shape_compute_module` cannot currently be deserialized in `compile()`, since mobile::Module has no IValue conversion.
- Processed arguments that are modified by `init()` must be kept as member variables. Any other processed arguments are passed through a dictionary, `handles`.

**nnapi_bind.cpp & nnapi_bind.h**: 
Created a header file for `nnapi_bind.cpp`, so that it's NnapiCompilation class can be used by `nnapi_backend_lib.cpp`.
**test_backend_nnapi.py**:
Enabled execution testing.

### Test Plan:
Tested on devserver.
1. Load and unpack a special devserver build of NNAPI: `jf download GICWmAAzUR0eo20TAPasVts8ObhobsIXAAAz --file "nnapi-host-linux.tar.xz"`
2. `export LIBNEURALNETWORKS_PATH=/path/to/libneuralnetworks.so`
3. Run unittests: `python test/test_jit.py TestNnapiBackend` and `python test/test_nnapi.py`

TODO: test with lite interpreter runtime

Reviewers:

Subscribers:

Tasks: T91991928

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]

Amyh11325 commented:
Added comments and removed unnecessary headers. Thank you for the comments!

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29944873).

Commit:
Update on "[1/N] Nnapi backend compile and execute"


### Summary:
Added Android NNAPI delegate implementation of runtime initialization (compilation) and execution.
The delegate's preprocess step was [previously implemented](https://github.com/pytorch/pytorch/pull/62225). Now, the rest of the delegate, which implements client-side execution, is added.

**nnapi_backend_lib.cpp**:
Implementation of delegate's compile and execute. 
`execute()` is essentially a C++ implementation of [`NnapiModule`](https://github.com/pytorch/pytorch/blob/master/torch/backends/_nnapi/prepare.py), which wraps an NNAPI Compilation and handles preparation of weights, inputs, and outputs.
- Any steps that can be done before execution are moved to `compile()`.
    - `init()` cannot be moved to `compile()` because it requires real inputs for dynamic shaping.
    - `shape_compute_module` cannot currently be deserialized in `compile()`, since mobile::Module has no IValue conversion.
- Processed arguments that are modified by `init()` must be kept as member variables. Any other processed arguments are passed through a dictionary, `handles`.

**nnapi_bind.cpp & nnapi_bind.h**: 
Created a header file for `nnapi_bind.cpp`, so that it's NnapiCompilation class can be used by `nnapi_backend_lib.cpp`.
**test_backend_nnapi.py**:
Enabled execution testing.

### Test Plan:
Tested on devserver.
1. Load and unpack a special devserver build of NNAPI: `jf download GICWmAAzUR0eo20TAPasVts8ObhobsIXAAAz --file "nnapi-host-linux.tar.xz"`
2. `export LIBNEURALNETWORKS_PATH=/path/to/libneuralnetworks.so`
3. Run unittests: `python test/test_jit.py TestNnapiBackend` and `python test/test_nnapi.py`

TODO: test with lite interpreter runtime

Reviewers:

Subscribers:

Tasks: T91991928

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]

Amyh11325 commented:
Export from internal to make sure everything matches.

Commit:
Update on "[1/N] Nnapi backend compile and execute"


### Summary:
Added Android NNAPI delegate implementation of runtime initialization (compilation) and execution.
The delegate's preprocess step was [previously implemented](https://github.com/pytorch/pytorch/pull/62225). Now, the rest of the delegate, which implements client-side execution, is added.

**nnapi_backend_lib.cpp**:
Implementation of delegate's compile and execute. 
`execute()` is essentially a C++ implementation of [`NnapiModule`](https://github.com/pytorch/pytorch/blob/master/torch/backends/_nnapi/prepare.py), which wraps an NNAPI Compilation and handles preparation of weights, inputs, and outputs.
- Any steps that can be done before execution are moved to `compile()`.
    - `init()` cannot be moved to `compile()` because it requires real inputs for dynamic shaping.
    - `shape_compute_module` cannot currently be deserialized in `compile()`, since mobile::Module has no IValue conversion.
- Processed arguments that are modified by `init()` must be kept as member variables. Any other processed arguments are passed through a dictionary, `handles`.

**nnapi_bind.cpp & nnapi_bind.h**: 
Created a header file for `nnapi_bind.cpp`, so that it's NnapiCompilation class can be used by `nnapi_backend_lib.cpp`.
**test_backend_nnapi.py**:
Enabled execution testing.

### Test Plan:
Tested on devserver.
1. Load and unpack a special devserver build of NNAPI: `jf download GICWmAAzUR0eo20TAPasVts8ObhobsIXAAAz --file "nnapi-host-linux.tar.xz"`
2. `export LIBNEURALNETWORKS_PATH=/path/to/libneuralnetworks.so`
3. Run unittests: `python test/test_jit.py TestNnapiBackend` and `python test/test_nnapi.py`

TODO: test with lite interpreter runtime

Reviewers:

Subscribers:

Tasks: T91991928

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]

Commit:
Update on "[1/N] Nnapi backend compile and execute"


### Summary:
Added Android NNAPI delegate implementation of runtime initialization (compilation) and execution.
The delegate's preprocess step was [previously implemented](https://github.com/pytorch/pytorch/pull/62225). Now, the rest of the delegate, which implements client-side execution, is added.

**nnapi_backend_lib.cpp**:
Implementation of delegate's compile and execute. 
`execute()` is essentially a C++ implementation of [`NnapiModule`](https://github.com/pytorch/pytorch/blob/master/torch/backends/_nnapi/prepare.py), which wraps an NNAPI Compilation and handles preparation of weights, inputs, and outputs.
- Any steps that can be done before execution are moved to `compile()`.
    - `init()` cannot be moved to `compile()` because it requires real inputs for dynamic shaping.
    - `shape_compute_module` cannot currently be deserialized in `compile()`, since mobile::Module has no IValue conversion.
- Processed arguments that are modified by `init()` must be kept as member variables. Any other processed arguments are passed through a dictionary, `handles`.

**nnapi_bind.cpp & nnapi_bind.h**: 
Created a header file for `nnapi_bind.cpp`, so that it's NnapiCompilation class can be used by `nnapi_backend_lib.cpp`.
**test_backend_nnapi.py**:
Enabled execution testing.

### Test Plan:
Tested on devserver.
1. Load and unpack a special devserver build of NNAPI: `jf download GICWmAAzUR0eo20TAPasVts8ObhobsIXAAAz --file "nnapi-host-linux.tar.xz"`
2. `export LIBNEURALNETWORKS_PATH=/path/to/libneuralnetworks.so`
3. Run unittests: `python test/test_jit.py TestNnapiBackend` and `python test/test_nnapi.py`

TODO: test with lite interpreter runtime

Reviewers:

Subscribers:

Tasks: T91991928

Tags:

Differential Revision: [D29944873](https://our.internmc.facebook.com/intern/diff/D29944873)

[ghstack-poisoned]
